src/Draco.Compiler/Internal/Syntax/Formatting/DracoFormatter.cs

```diff
@@ -0,0 +1,440 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Draco.Compiler.Api.Syntax;
+
+namespace Draco.Compiler.Internal.Syntax.Formatting;
+
+internal sealed class DracoFormatter : Api.Syntax.SyntaxVisitor
+{
+    private readonly FormatterSettings settings;
+    private FormatterEngine formatter = null!;
+    private DracoFormatter(FormatterSettings settings)
+    {
+        this.settings = settings;
+    }
+
+    /// <summary>
+    /// Formats the given syntax tree.
+    /// </summary>
+    /// <param name="tree">The syntax tree to format.</param>
+    /// <param name="settings">The formatter settings to use.</param>
+    /// <returns>The formatted tree.</returns>
+    public static string Format(SyntaxTree tree, FormatterSettings? settings = null)
+    {
+        settings ??= FormatterSettings.Default;
+
+        var formatter = new DracoFormatter(settings);
+        tree.Root.Accept(formatter);
+
+        var metadatas = formatter.formatter.TokensMetadata;
```

Comment:
If you cache the formatter anyway, why not take it in the ctor?

Reply:
wdym that I cache it ?  
What would it bring ?

I think the API is indeed weird currently, I'll think about it and push changes to improve the formatter API.
Currently it's just calling a magic static method from somewhere on some field.